mediatek: filogic: add ZyXEL EX5700 (Telenor) support

bmork opened the pull request:
Add support for the Telenor branded version of the ZyXEL EX5700. This comes with a modified U-Boot requiring device specific workarounds.

bmork committed:
kernel: ubootenv-nvram: driver for RAM backed environments

The vendor U-Boot implementaion on Telenor branded ZyXEL EX5700
devices does not store its environment on flash. It is instead
kept in a memory region.  This is persistent over reboots, but
not over power cycling.

The dual partition failsafe system used by the vendor U-Boot
requires the OS to modify a variable in this memory environment.
This driver allows the ordinary uboot-envtools to access a
memory region like it was a partition on NOR flash.

The specific vendor U-Boot adds a "no-map" /reserved-memory
section and a top level /ubootenv node pointing to the memory
environment.  The driver uses this device specific fact to
locate the region.  The matching and probing code will likely
have to be adjusted for any other devices to be supported.

Example partial device tree:

 / {
    ..
    ubootenv {
        memory-region = <&uenv>;
        compatible = "ubootenv";
    };
    ..
    reserved-memory {
        ..
        uenv: ubootenv@7ffe8000 {
            no-map;
            reg = <0 0x7ffe8000 0 0x4000>;
        };

Signed-off-by: Bjørn Mork <bjorn@mork.no>

bmork committed:
filogic: support Telenor branded ZyXEL EX5700

Telenor quirks
--------------
The operator specific firmware running on the Telenor branded
ZyXEL EX5700 includes U-Boot modifications affecting the OpenWrt
installation.

Notable changes to U-Boot include
- environment is stored in RAM and reset to defaults when power
  cycled
- dual partition scheme with "nomimal" or "rescue" systems, falling
  back to "rescue" unless the OS signals success in 3 attempts
- several runtime additions to the device-tree

Some of these modifications have side effects requiring workarounds
- U-Boot modifies /chosen/bootargs in an unsafe manner, and will crash
  unless this node exists
- U-Boot verifies that the selected rootfs UBI volume exists, and
  refuses to boot if it doesn't. The chosen "rootfs" volume must contain
  a squashfs signature even for tftp or initramfs booting.
- U-Boot parses the "factoryparams" UBI volume, setting the "ethaddr"
  variable to the label mac.  But "factoryparams" does not always
  exist.  Instead there is a "RIP" volume containing all the factory
  data.  Copying the "RIP" volume to  "factoryparams" will fix this

Hardware
--------
SOC:   MediaTek MT7986
RAM:   1GB DDR4
FLASH: 512MB SPI-NAND (Mikron xxx)
WIFI:  Mediatek MT7986 802.11ax 5 GHz
       Mediatek MT7916 DBDC 802.11ax 2.4 + 6 GHz
ETH:   MediaTek MT7531 Switch + SoC
       3 x builtin 1G phy (lan1, lan2, lan3)
       2 x MaxLinear GPY211C 2.5 N-Base-T phy (lan4, wan)
USB:   1 x USB 3.2 Enhanced SuperSpeed port
UART:  3V3 115200 8N1 (Pinout: GND KEY RX TX VCC)

Installation
------------
1. Download the OpenWrt initramfs image. Copy the image to a TFTP server
   reachable at 192.168.1.2/24. Rename the image to C0A80101.img.

2. Connect the TFTP server to lan1, lan2 or lan3. Connect to the serial
   console, Interrupt the autoboot process by pressing ESC when prompted.

3. Download and boot the OpenWrt initramfs image.

   $ env set uboot_bootcount 0
   $ env set firmware nominal
   $ tftpboot
   $ bootm

4. Wait for OpenWrt to boot. Transfer the sysupgrade image to the device
   using scp and install using sysupgrade.

   $ sysupgrade -n <path-to-sysupgrade.bin>

Missing features
----------------

- The "lan1", "lan2" and "lan3" port LEDs are driven by the switch but
  OpenWrt does not correctly configure the output.
- The "lan4" and "wan" port LEDs are driven by the GPH211C phys and
  not configured by OpenWrt.

Signed-off-by: Bjørn Mork <bjorn@mork.no>

aiamadeus commented:
Hi, the "mediatek,eth-mac label" patch has been rejected in #11435.

bmork commented:
> Hi, the "mediatek,eth-mac label" patch has been rejected in #11435.

Thanks for the info.  Makes absolutely no sense given that the patch already is in OpenWrt....  But in principle, I agree with the idea behind it - follow upstream as closely as possible.

AFAICS, "label" is documented in Documentation/devicetree/bindings/net/ethernet-controller.yaml and implemented in at least the DSA subsystem.  Allowing that, which could just as well have been done in userspace, but rejecting it for other mac drivers is inconsistent.

I'll remove this patch here, and replace it with some hacky userspace solution.

Thinking about trying upstream again. Too bad it's hard (impossible?) to implement this in some common infrastructure.  Looks like it has to be done in every driver?

> > Hardware
> > RAM:   1GB DDR3
> 
> Should it be 1GB DDR4 here?


Probably.  Will fix.

Thanks for the review

csharper2005 commented:
```
WIFI:  Mediatek MT7986 802.11ax 5 GHz
       Mediatek MT7916 DBDC 802.11ax 2.4 + 6 GHz
```
Megadevice. 0_o

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -185,3 +185,19 @@ ifneq ($(CONFIG_TARGET_ROOTFS_INITRAMFS),)
+  DEVICE_DTS := mt7986a-zyxel-ex5700-telenor
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-mt7916-firmware kmod-ubootenv-nvram kmod-usb3 kmod-mt7986-firmware mt7986-wo-firmware
+  KERNEL_LOADADDR := 0x48000000
```
This is present in device/default. So it can be safely dropped from here.

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -185,3 +185,19 @@ ifneq ($(CONFIG_TARGET_ROOTFS_INITRAMFS),)
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  IMAGE_SIZE := 65536k
+  KERNEL_IN_UBI := 1
```
Why do you need `KERNEL_IN_UBI`? I didn't find any ubinized images in your recipes.

bmork replied:
thanks.  Can't explain.  Plrobably stupid cutnpaste.  Will fix

1980490718 replied:
> thanks. Can't explain. Plrobably stupid cutnpaste. Will fix
  BLOCKSIZE := 2048k
  PAGESIZE := 4096
  IMAGE_SIZE := 65536k
  KERNEL_IN_UBI := 1

1980490718 replied:
> thanks. Can't explain. Plrobably stupid cutnpaste. Will fix
maybe BLOCKSIZE := 256k & PAGESIZE := 4096
@bmork @hauke

csharper2005 commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5700-telenor.dts:
```diff
@@ -0,0 +1,367 @@
+
+		port@6 {
+			reg = <6>;
+			label = "cpu";
```
`cpu` label can be dropped

bmork replied:
will fix

csharper2005 commented on scripts/mkits.sh:
```diff
@@ -108,10 +108,10 @@ if [ -n "${DTB}" ]; then
 			arch = \"${ARCH}\";
 			compression = \"none\";
-			hash@1 {
+			hash${REFERENCE_CHAR}1 {
```
Where are you declaring `REFERENCE_CHAR`?

bmork committed:
mediatek: filogic: set wan label in preinit

Implement the functionality of
target/linux/ramips/patches-5.15/700-net-ethernet-mediatek-support-net-labels.patch
in userspace, since the driver patch has been rejected as a generic solution:
https://github.com/openwrt/openwrt/pull/11435

Signed-off-by: Bjørn Mork <bjorn@mork.no>

bmork commented:
Pushed an updated branch with all comments handled.  Changes:

- rebased to current master
- fixed RAM description
- ethernet driver label patch replaced by userspace solution
- dropped unneccessary device vars
- dropped "cpu" label from dts

bmork commented:
rebased onto master, resolving the conflicts caused by the recently added EX5600 sister device

bmork commented:
updated to account for commit 7a0ec001ff79 ("mediatek: sync MT7986 device trees with upstream") and rebased onto current main

bmork commented:
- rebased onto main after the mediatek v6.1 kernel support was merged
- fixed up a number of checkpatch whitespace warnings
- run-tested with both v5.15 and v6.1 kernels

bmork commented:
Yet another rebase on main, dropping the mkits patch which was merged as commit 2b133ab19cd5 ("scripts: use sep-char for hash nodes")

bmork commented:
rebased to resolve conflicts after the NWA50AX Pro addition
